### DISCUSSION TITLE
New folder image replacement and removal

### DISCUSSION THREAD

**Discussion by JorjMcKie** [pymupdf/PyMuPDF-Utilities#35] [Announcements]
There is a new folder on [image replacement](https://github.com/pymupdf/PyMuPDF-Utilities/tree/master/image-replacement), which contains scripts and documentation to explain and compare available options.

**Comment by foranuj** [comment]
Thanks for this, very useful scripts.

**Reply by JorjMcKie** [reply]
Thank you!
Please note that both, image replacement and deletion are now official methods `Page.replace_image()`/ `Page.delete_image()` in the newest version 1.21.1 that came out a few days ago.
I will go by and leave a comment in this folder.

**Reply by foranuj** [reply]
What about the transformations on the image? Are those handled automatically as well? I have a photo collage and when I use the replace to put a whole image in a rectangle, I see that the entire image is replaced, and squished onto the page in that original small rectangle...

**Reply by JorjMcKie** [reply]
> What about the transformations on the image? Are those handled automatically as well? I have a photo collage and when I use the replace to put a whole image in a rectangle, I see that the entire image is replaced, and squished onto the page in that original small rectangle...

Yes, and that is inevitable. This technique shouldn't be over-estimated: what technically happens is this:
1. You have an image at xref 46, referred to in the page object definition via `/Resources<</XObject<</fzImg0 46 0 R>>>>`.
2. The page's appearance source somewhere contains commands like these:
```
q
297.66 0 0 339.69856 200.00002 261.59074 cm
/fzImg0 Do
Q
```
The "cm" command represents a matrix which takes care of mapping the image to the desired page area. The "Do" command actually invokes the image identified by name fzImg0, which obviously points to xref 46.

By replacing the image originally found under xref 46, **_nothing else changes._** The same matrix is applied, the name fzImg0 is not changed, etc.

So if your new image has different width / height values then bad luck! Especially ugly the result if the width:height ratio is not the same ...

**Reply by JorjMcKie** [reply]
In a situation as you indicate I would use the normal `page.insert_image()` method. Potentially after removing any previous image.
